docs/source/advanced.rst

```diff
@@ -128,6 +128,43 @@ offer an easy way to compile and tune compilation parameters. Examples:
 In all cases, returning an un-compiled model is equivalent to
 calling ``model.compile(**compile_kwargs)`` within ``model_build_fn``.
 
+.. _loss-selection:
+
+Loss selection
+++++++++++++++
+
+If you do not explicitly define a loss, SciKeras attempts to find a loss
+that matches the type of target (see :py:func:`sklearn.utils.multiclass.type_of_target`).
+
+For guidance selecting losses in Keras, please see Jason Brownlee's
+excellent article `How to Choose Loss Functions When Training Deep Learning Neural Networks`_
+as well as `Keras Losses docs`_.
+
+Default losses are selected as follows:
+
+Classification
+..............
+
++-----------+-----------+----------+---------------------------------+
+| # outputs | # classes | encoding | loss                            |
```

Comment:
I'm confused by this table. Let's say I have two classes, one "output," and I don't know my "encoding" (I'm not sure a naive user would know what that means). What loss is chosen?

Maybe it'd be simpler to say "KerasClassifier has `loss="sparse_categorical_crossentropy"` by default. It works for one dimensional labels like `est.fit(X, [2, 3, 4, 5])`). If you have binary labels like `y=[-1, 1, -1, -1]`, specify `binary_crossentropy`. If you have one-hot encoded labels, use LOSS."

Reply:
How about:

> KerasClassifier will automatically determine an appropriate loss function for binary (`[0, 1, 0]`/`["car", "bike", "car"]`) or multiclass (`[1, 2, 3, 4]`/`["person", "car", "pear", "tree"]`) one-dimensional targets. For other types of target, you must explicitly specify the loss. If your target is one-hot encoded, you probably want to use `"categorical_crossentropy"`.

Reply:
I'm only for using `loss="auto"` if there are simple and easy-to-follow rules. Setting a fixed value of `loss="sparse_categorical_crossentropy"` is a really simple rule.

I almost prefer this documentation:

> KerasClassifier has `loss="sparse_categorical_crossentropy"` by default. This assumes that the model has C outputs neurons to classify C classes. It's intended to be used like this:
>
> ``` python
> def build_model():
>     ...
>     model.add(output_layer_C_neurons)
>     return model
> 
> est = KerasClassifier(model=build_model)
> est.fit(X, [0, 1, 2, 0, 1, 2])
> ```
> 
> If you have one-hot encoded targets, manually 
>
> ``` python
> from sklearn.datasets import OneHotEncoder
> est = KerasClassifier(model=build_model, loss="categorical_crossentropy")
> y = OneHotEncoder([0, 1, 2, 0, 1, 2])
> y = [[1, 0, 0], [0, 1, 0], [0, 0, 1], [1, 0, 0], [0, 1, 0], [0, 0, 1]]  # or this
> est.fit(X, y)
> ```

Reply:
I think we tried this before. I don't remember the conclusion of those discussions (although I can dig it up), but off the top of my head I think the biggest issue is that new users will copy an example model from a tutorial, many of which do binary classification using a single neuron, or other incompatible architectures.

Do you think we can just introspect the model and check if the number of neurons matches the number of classes (and that it is a single-output problem) and raise an error (or maybe a warning) to rescue users from facing whatever cryptic error TF would throw?

Reply:
> I don't remember the conclusion of those discussions (although I can dig it up

I recall introspecting the model to see what loss value should be used, but trying to abstract too much away from the user (and plus it got too complicated).

I think the new loss for `KerasClassifier` is better: it's very simple and recommend changes if common mistakes are made (eventually; see below).

> introspect the model and ... raise an error (or maybe a warning)

Yeah, I had the same idea. If I were developing this library, I think I'd have `loss="sparse_categorical_crossentropy"` with clear documentation ("have model return one neuron for each output, likely with softmax activation"). I would catch these use cases:

* 1 output neuron and `loss != "binary_crossentropy"`.
* target one-hot encoded (and tell to set loss="categorical_crossentropy"`).

I think both of these should be exceptions. If so, I'd make it clear how to clear how to adapt to BaseWrapper.

> copy an example model from a tutorial, many of which do binary classification using a single neuron

I think a clear documentation note would resolve this, especially with good error catching.

<details>
<summary>keras.io examples</summary>

* One neuron: [3D CT scans][3], [text classification][7]
* Dynamic choice: [image classification][1], 
* ≥ 2 neurons: [simple mnist][2], [EfficientNet][4], [ViT][5], [Perceiver][6], [text w/ switch][8], [text w/ transformer][9a]

</summary>

[1]:https://keras.io/examples/vision/image_classification_from_scratch/#build-a-model
[2]:https://keras.io/examples/vision/mnist_convnet/#build-the-model
[3]:https://keras.io/examples/vision/3D_image_classification/#define-a-3d-convolutional-neural-network
[4]:https://keras.io/examples/vision/image_classification_efficientnet_fine_tuning/#training-a-model-from-scratch
[5]:https://keras.io/examples/vision/image_classification_with_vision_transformer/#build-the-vit-model
[6]:https://keras.io/examples/vision/perceiver_image_classification/#prepare-the-data
[7]:https://keras.io/examples/nlp/text_classification_from_scratch/#build-a-model
[8]:https://keras.io/examples/nlp/text_classification_with_switch_transformer/#implement-the-classifier
[9a]:https://keras.io/examples/nlp/text_classification_with_transformer/#create-classifier-model-using-transformer-layer

Reply:
I think https://github.com/adriangb/scikeras/pull/210#discussion_r627916340 is at at least worth exploring (again).

I'll open a new PR to test out https://github.com/adriangb/scikeras/pull/210#discussion_r627916340 and to avoid loosing the changes here into the git history, and also because the changes are going to be pretty unrelated.

Thank you for following up on this PR 😄